commit 50296b42e3e2
Author: Andrea Marchesini <amarchesini@mozilla.com>
Date:   Thu Oct 19 19:06:18 2017 +0200

    Bug 1407936 - Cleaning up QuotaManagar data only when ServiceWorkers are correctly unregistered. r=bkelly, a=ritu
    
    --HG--
    extra : source : 3624656ebe465779d8ec8fa4103d477a0e725165
    extra : histedit_source : 4a0d0011a098a070d2da7c5288d51c2cbed01e1c
---
 browser/base/content/sanitize.js                | 19 ++++++++++++++++---
 dom/interfaces/base/nsIServiceWorkerManager.idl |  3 ---
 2 files changed, 16 insertions(+), 6 deletions(-)

diff --git browser/base/content/sanitize.js browser/base/content/sanitize.js
index 25c56c47bc2e..27e1343bc89b 100644
--- browser/base/content/sanitize.js
+++ browser/base/content/sanitize.js
@@ -297,15 +297,28 @@ Sanitizer.prototype = {
         Services.obs.notifyObservers(null, "extension:purge-localStorage");
 
         // ServiceWorkers
+        let promises = [];
         let serviceWorkers = serviceWorkerManager.getAllRegistrations();
         for (let i = 0; i < serviceWorkers.length; i++) {
           let sw = serviceWorkers.queryElementAt(i, Ci.nsIServiceWorkerRegistrationInfo);
-          let host = sw.principal.URI.host;
-          serviceWorkerManager.removeAndPropagate(host);
+
+          promises.push(new Promise(resolve => {
+            let unregisterCallback = {
+              unregisterSucceeded: () => { resolve(true); },
+              // We don't care about failures.
+              unregisterFailed: () => { resolve(true); },
+              QueryInterface: XPCOMUtils.generateQI(
+                [Ci.nsIServiceWorkerUnregisterCallback])
+            };
+
+            serviceWorkerManager.propagateUnregister(sw.principal, unregisterCallback, sw.scope);
+          }));
         }
 
+        await Promise.all(promises);
+
         // QuotaManager
-        let promises = [];
+        promises = [];
         await new Promise(resolve => {
           quotaManagerService.getUsage(request => {
             if (request.resultCode != Cr.NS_OK) {
diff --git dom/interfaces/base/nsIServiceWorkerManager.idl dom/interfaces/base/nsIServiceWorkerManager.idl
index 6b4505f7e185..42e38f4afceb 100644
--- dom/interfaces/base/nsIServiceWorkerManager.idl
+++ dom/interfaces/base/nsIServiceWorkerManager.idl
@@ -186,16 +186,13 @@ interface nsIServiceWorkerManager : nsISupports
   // Testing
   DOMString getScopeForUrl(in nsIPrincipal aPrincipal, in DOMString aPath);
 
-  // Note: This is meant to be used only by about:serviceworkers.
   // It returns an array of nsIServiceWorkerRegistrationInfos.
   nsIArray getAllRegistrations();
 
-  // Note: This is meant to be used only by about:serviceworkers.
   // It calls softUpdate() for each child process.
   [implicit_jscontext] void propagateSoftUpdate(in jsval aOriginAttributes,
                                                 in DOMString aScope);
 
-  // Note: This is meant to be used only by about:serviceworkers.
   // It calls unregister() in each child process. The callback is used to
   // inform when unregister() is completed on the current process.
   void propagateUnregister(in nsIPrincipal aPrincipal,
